src/borrowing.rs

```diff
@@ -0,0 +1,109 @@
+//! Non-allcating, safe, statically allocated wrapper for `DiatomicWaker`.
+//!
+//! See the [crate-level documentation](crate) for usage.
+
+use std::task::Waker;
+
+use crate::primitives::DiatomicWaker;
+use crate::primitives::WaitUntil;
+
+/// A safe, statically allocated `DiatomicWaker`
+///
+/// See the [crate-level documentation](crate) for usage.
+#[derive(Default, Debug)]
+pub struct BorrowingDiatomicWaker {
```

Comment:
I wouldn't mind adding `split` to `DiatomicWaker` directly, do you see any problem with that? The `unsafe` methods cannot be used while the borrowing `WakeSource` and `WakeSink` are active, so this wouldn't introduce additional footguns.